fog/overseer/server/src/worker.rs

```diff
@@ -0,0 +1,546 @@
+    ///             (i) Disarm
+    ///             (ii) Send an alert (todo).
+    fn perform_automatic_failover(
+        &self,
+        ingest_summary_node_mappings: Vec<IngestSummaryNodeMapping>,
+    ) -> Result<(), OverseerError> {
+        let inactive_outstanding_keys: Vec<CompressedRistrettoPublic> =
+            self.get_inactive_outstanding_keys()?;
+
+        match inactive_outstanding_keys.len() {
+            0 => log::info!(self.logger, "Found 0 outstanding keys."),
+            1 => log::info!(self.logger, "Found 1 outstanding key."),
+            _ => {
+                log::error!(self.logger, "Found multiple outstanding keys. This requires manual intervention. Disarming.");
+                self.stop_requested.store(true, Ordering::SeqCst);
+                return Err(OverseerError::MultipleOutstandingKeys("Multiple outstanding keys found. This is unexpected and requires manual intervention. As such, we've disarmed overseer. Take action and then rearm.".to_string()));
+            }
+        }
+
+        if let Some(inactive_outstanding_key) = inactive_outstanding_keys.get(0) {
+            match self.try_to_activate_node_for_inactive_outstanding_key(
+                inactive_outstanding_key,
+                &ingest_summary_node_mappings,
+            ) {
+                Ok(_) => {
+                    log::info!(
+                        self.logger,
+                        "A node was activated, so there is no need to report the key as lost."
+                    );
+                    return Ok(());
+                }
+                Err(err) => log::error!(
+                    self.logger,
+                    "There was an issue activating a node for the outstanding key: {:?}",
+                    err
+                ),
+            }
+        }
+
+        // If there's an outstanding key, we need to report it as lost.
+        if let Some(inactive_outstanding_key) = inactive_outstanding_keys.get(0) {
+            self.report_lost_ingress_key(*inactive_outstanding_key)?;
+        }
+
+        // Regardless of whether or not there's an outstanding key, we need
+        // to set new keys on a node and then activate it.
+        let activated_node_index = self.set_new_key_on_a_node()?;
+        self.activate_a_node(activated_node_index)?;
+
+        Ok(())
+    }
+
+    fn get_inactive_outstanding_keys(
+        &self,
+    ) -> Result<Vec<CompressedRistrettoPublic>, OverseerError> {
+        // An outanding key is one that Fog Ingest is still obligated to be
+        // scanning blocks with on behalf of users.
+        let outstanding_keys_filters = IngressPublicKeyRecordFilters {
+            // A lost key is not outstanding because it will never again be used
+            // to scan blocks.
+            should_include_lost_keys: false,
+            // A retired key can be outstanding if public expiry is greater than
+            // its last scanned block.
+            should_include_retired_keys: true,
```

Comment:
The comment on should include retired keys and should_only_include_unexpired_keys is a bit confusing. Could we clarify this a little bit? How are these 2 different?

Reply:
The two filters are different in that the `should_include_retired_keys` set to `true` includes keys that have been marked retired in the recovery DB (setting it to `false` discards any key that is marked retired). The second filter, `should_only_include_unexpired_keys` discards any key that is expired- an expired key is one in which it's pubkey_expiry (i.e. the last block for which the key must be used to scan) is <= the last scanned block. 

 We need retired keys because they can still be outstanding, meaning they might still have work to do because their pubkey_expiry (i.e. the last block for which they are responsible for scanning) might be greater than the last scanned block.

 I added to the `should_include_retired_keys` comment to try and clarify this point. Let me know if that helps or if you'd like to see anything else written in this comment.